lib/train/plugins/base_connection.rb

```diff
@@ -52,39 +74,37 @@ def local?
       false
     end
 
-    # Execute a command using this connection.
-    #
-    # @param command [String] command string to execute
-    # @return [CommandResult] contains the result of running the command
-    def run_command(_command)
-      fail Train::ClientError, "#{self.class} does not implement #run_command()"
-    end
-
     # Get information on the operating system which this transport connects to.
     #
     # @return [Platform] system information
     def platform
       @platform ||= Train::Platforms::Detect.scan(self)
     end
-
     # we need to keep os as a method for backwards compatibility with inspec
     alias os platform
 
-    # Interact with files on the target. Read, write, and get metadata
-    # from files via the transport.
-    #
-    # @param [String] path which is being inspected
-    # @return [FileCommon] file object that allows for interaction
-    def file(_path, *_args)
-      fail Train::ClientError, "#{self.class} does not implement #file(...)"
+    # This is the main command call for all connections. This will call the private
+    # run_command_via_connection on the connection with optional caching
+    def run_command(cmd)
+      return @cache[:command][cmd] ||= run_command_via_connection(cmd) if cache_enabled?(:command)
```

Comment:
I think it might be a bit cleaner to read if it was structured to do the non-cache enabled stuff first:

```ruby
return run_command_via_connection(cmd) unless cache_enabled?(:command)

@cache[:command][cmd] ||= run_command_via_connection(cmd)
```

The `return` combined with the `||=` is hanging me up a bit. Feels like too much happening in one conditional statement.

Reply:
Uh, I like that one. Thank you @adamleff